### PR TITLE
ct: Use more volatile

### PR DESCRIPTION
**Pull request by real-or-random**
The first commit adds a `volatile` to fix a constant-time issue in the ECDH code with GCC 13.1. The second commit adds a few more `volatile`s to conditional "flag" variables. 

All of these cases were missed in https://github.com/bitcoin-core/secp256k1/pull/1257. I haven't looked at the entire code base, but I think this should cover most (all?) of it. All of this is of course a bit arbitrary, but I think a good rule of thumb is to use `volatile` whenever we turn a secret boolean variable into a bitmask.

### CODE REVIEW

File: src/scalar_4x64_impl.h [hunk 1]
```diff
@@ -189,7 +190,8 @@ static int secp256k1_scalar_is_high(const secp256k1_scalar *a) {
 static int secp256k1_scalar_cond_negate(secp256k1_scalar *r, int flag) {
     /* If we are flag = 0, mask = 00...00 and this is a no-op;
      * if we are flag = 1, mask = 11...11 and this is identical to secp256k1_scalar_negate */
-    uint64_t mask = !flag - 1;
+    volatile int vflag = flag;
+    uint64_t mask = 0 - vflag;
```
Comment (real-or-random) [inline src/scalar_4x64_impl.h:194]: nb: this is intentionally `0 - vflag` instead of `-vflag` to avoid warnings in MSVC, see #1293

**Review by jonasnick — COMMENTED** [review]
Are we not missing the conditional operations in `divststeps_30` and `divsteps_59`?

File: src/scalar_4x64_impl.h (hunk 1 above)
Comment (real-or-random) [reply src/scalar_4x64_impl.h]: I've changed my mind. This should not be the concern of this PR.

**Comment by jonasnick** [comment]
ACK 17fa21733aae97bf671fede3ce528c7a3b2f5f14

I have not benchmarked performance

**Comment by real-or-random** [comment]
For posterity, here's the full output of the constant-time tests.

```
==31021== Memcheck, a memory error detector
==31021== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==31021== Using Valgrind-3.21.0 and LibVEX; rerun with -h for copyright info
==31021== Command: /home/tim/bs/dev/secp256k1/.libs/ctime_tests
==31021== 
==31021== Conditional jump or move depends on uninitialised value(s)
==31021==    at 0x4856355: secp256k1_u128_accum_u64 (int128_native_impl.h:20)
==31021==    by 0x4856355: secp256k1_scalar_cond_negate (scalar_4x64_impl.h:196)
==31021==    by 0x485DD3F: secp256k1_ecdsa_sig_sign (ecdsa_impl.h:305)
==31021==    by 0x485DD3F: secp256k1_ecdsa_sign_inner (secp256k1.c:532)
==31021==    by 0x486165A: secp256k1_ecdsa_sign (secp256k1.c:567)
==31021==    by 0x10974D: run_tests (ctime_tests.c:98)
==31021==    by 0x109252: main (ctime_tests.c:53)
==31021== 
==31021== Conditional jump or move depends on uninitialised value(s)
==31021==    at 0x4856371: secp256k1_u128_accum_u64 (int128_native_impl.h:20)
==31021==    by 0x4856371: secp256k1_scalar_cond_negate (scalar_4x64_impl.h:199)
==31021==    by 0x485DD3F: secp256k1_ecdsa_sig_sign (ecdsa_impl.h:305)
==31021==    by 0x485DD3F: secp256k1_ecdsa_sign_inner (secp256k1.c:532)
==31021==    by 0x486165A: secp256k1_ecdsa_sign (secp256k1.c:567)
==31021==    by 0x10974D: run_tests (ctime_tests.c:98)
==31021==    by 0x109252: main (ctime_tests.c:53)
==31021== 
==31021== Conditional jump or move depends on uninitialised value(s)
==31021==    at 0x48563A8: secp256k1_u128_accum_u64 (int128_native_impl.h:20)
==31021==    by 0x48563A8: secp256k1_scalar_cond_negate (scalar_4x64_impl.h:202)
==31021==    by 0x485DD3F: secp256k1_ecdsa_sig_sign (ecdsa_impl.h:305)
==31021==    by 0x485DD3F: secp256k1_ecdsa_sign_inner (secp256k1.c:532)
==31021==    by 0x486165A: secp256k1_ecdsa_sign (secp256k1.c:567)
==31021==    by 0x10974D: run_tests (ctime_tests.c:98)
==31021==    by 0x109252: main (ctime_tests.c:53)
==31021== 
==31021== Conditional jump or move depends on uninitialised value(s)
==31021==    at 0x4856355: secp256k1_u128_accum_u64 (int128_native_impl.h:20)
==31021==    by 0x4856355: secp256k1_scalar_cond_negate (scalar_4x64_impl.h:196)
==31021==    by 0x485B304: secp256k1_wnaf_const.constprop.0 (ecmult_const_impl.h:99)
==31021==    by 0x4862657: secp256k1_ecmult_const (ecmult_const_impl.h:155)
==31021==    by 0x4862657: secp256k1_ecdh (main_impl.h:53)
==31021==    by 0x1098F0: run_tests (ctime_tests.c:107)
==31021==    by 0x109252: main (ctime_tests.c:53)
==31021== 
==31021== Conditional jump or move depends on uninitialised value(s)
==31021==    at 0x4856371: secp256k1_u128_accum_u64 (int128_native_impl.h:20)
==31021==    by 0x4856371: secp256k1_scalar_cond_negate (scalar_4x64_impl.h:199)
==31021==    by 0x485B304: secp256k1_wnaf_const.constprop.0 (ecmult_const_impl.h:99)
==31021==    by 0x4862657: secp256k1_ecmult_const (ecmult_const_impl.h:155)
==31021==    by 0x4862657: secp256k1_ecdh (main_impl.h:53)
==31021==    by 0x1098F0: run_tests (ctime_tests.c:107)
==31021==    by 0x109252: main (ctime_tests.c:53)
==31021== 
==31021== Conditional jump or move depends on uninitialised value(s)
==31021==    at 0x48563A8: secp256k1_u128_accum_u64 (int128_native_impl.h:20)
==31021==    by 0x48563A8: secp256k1_scalar_cond_negate (scalar_4x64_impl.h:202)
==31021==    by 0x485B304: secp256k1_wnaf_const.constprop.0 (ecmult_const_impl.h:99)
==31021==    by 0x4862657: secp256k1_ecmult_const (ecmult_const_impl.h:155)
==31021==    by 0x4862657: secp256k1_ecdh (main_impl.h:53)
==31021==    by 0x1098F0: run_tests (ctime_tests.c:107)
==31021==    by 0x109252: main (ctime_tests.c:53)
==31021== 
==31021== Conditional jump or move depends on uninitialised value(s)
==31021==    at 0x4856355: secp256k1_u128_accum_u64 (int128_native_impl.h:20)
==31021==    by 0x4856355: secp256k1_scalar_cond_negate (scalar_4x64_impl.h:196)
==31021==    by 0x485B304: secp256k1_wnaf_const.constprop.0 (ecmult_const_impl.h:99)
==31021==    by 0x4862683: secp256k1_ecmult_const (ecmult_const_impl.h:156)
==31021==    by 0x4862683: secp256k1_ecdh (main_impl.h:53)
==31021==    by 0x1098F0: run_tests (ctime_tests.c:107)
==31021==    by 0x109252: main (ctime_tests.c:53)
==31021== 
==31021== Conditional jump or move depends on uninitialised value(s)
==31021==    at 0x4856371: secp256k1_u128_accum_u64 (int128_native_impl.h:20)
==31021==    by 0x4856371: secp256k1_scalar_cond_negate (scalar_4x64_impl.h:199)
==31021==    by 0x485B304: secp256k1_wnaf_const.constprop.0 (ecmult_const_impl.h:99)
==31021==    by 0x4862683: secp256k1_ecmult_const (ecmult_const_impl.h:156)
==31021==    by 0x4862683: secp256k1_ecdh (main_impl.h:53)
==31021==    by 0x1098F0: run_tests (ctime_tests.c:107)
==31021==    by 0x109252: main (ctime_tests.c:53)
==31021== 
==31021== Conditional jump or move depends on uninitialised value(s)
==31021==    at 0x48563A8: secp256k1_u128_accum_u64 (int128_native_impl.h:20)
==31021==    by 0x48563A8: secp256k1_scalar_cond_negate (scalar_4x64_impl.h:202)
==31021==    by 0x485B304: secp256k1_wnaf_const.constprop.0 (ecmult_const_impl.h:99)
==31021==    by 0x4862683: secp256k1_ecmult_const (ecmult_const_impl.h:156)
==31021==    by 0x4862683: secp256k1_ecdh (main_impl.h:53)
==31021==    by 0x1098F0: run_tests (ctime_tests.c:107)
==31021==    by 0x109252: main (ctime_tests.c:53)
==31021== 
==31021== Conditional jump or move depends on uninitialised value(s)
==31021==    at 0x4856355: secp256k1_u128_accum_u64 (int128_native_impl.h:20)
==31021==    by 0x4856355: secp256k1_scalar_cond_negate (scalar_4x64_impl.h:196)
==31021==    by 0x485DD3F: secp256k1_ecdsa_sig_sign (ecdsa_impl.h:305)
==31021==    by 0x485DD3F: secp256k1_ecdsa_sign_inner (secp256k1.c:532)
==31021==    by 0x4863B36: secp256k1_ecdsa_sign_recoverable (main_impl.h:132)
==31021==    by 0x1099ED: run_tests (ctime_tests.c:115)
==31021==    by 0x109252: main (ctime_tests.c:53)
==31021== 
==31021== Conditional jump or move depends on uninitialised value(s)
==31021==    at 0x4856371: secp256k1_u128_accum_u64 (int128_native_impl.h:20)
==31021==    by 0x4856371: secp256k1_scalar_cond_negate (scalar_4x64_impl.h:199)
==31021==    by 0x485DD3F: secp256k1_ecdsa_sig_sign (ecdsa_impl.h:305)
==31021==    by 0x485DD3F: secp256k1_ecdsa_sign_inner (secp256k1.c:532)
==31021==    by 0x4863B36: secp256k1_ecdsa_sign_recoverable (main_impl.h:132)
==31021==    by 0x1099ED: run_tests (ctime_tests.c:115)
==31021==    by 0x109252: main (ctime_tests.c:53)
==31021== 
==31021== Conditional jump or move depends on uninitialised value(s)
==31021==    at 0x48563A8: secp256k1_u128_accum_u64 (int128_native_impl.h:20)
==31021==    by 0x48563A8: secp256k1_scalar_cond_negate (scalar_4x64_impl.h:202)
==31021==    by 0x485DD3F: secp256k1_ecdsa_sig_sign (ecdsa_impl.h:305)
==31021==    by 0x485DD3F: secp256k1_ecdsa_sign_inner (secp256k1.c:532)
==31021==    by 0x4863B36: secp256k1_ecdsa_sign_recoverable (main_impl.h:132)
==31021==    by 0x1099ED: run_tests (ctime_tests.c:115)
==31021==    by 0x109252: main (ctime_tests.c:53)
==31021== 
==31021== 
==31021== HEAP SUMMARY:
==31021==     in use at exit: 0 bytes in 0 blocks
==31021==   total heap usage: 1 allocs, 1 frees, 208 bytes allocated
==31021== 
==31021== All heap blocks were freed -- no leaks are possible
==31021== 
==31021== Use --track-origins=yes to see where uninitialised values come from
==31021== For lists of detected and suppressed errors, rerun with: -s
==31021== ERROR SUMMARY: 12 errors from 12 contexts (suppressed: 0 from 0)
```

The violation in ECDSA is a false positive because the affected scalar is public at this point, but the violation in ECDH is real.

**Pull request merged**